File: py/generateParentheses.py
"""
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
"""


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        if not n:
            return res
        self.helper(0, 0, n, '', res)
        return res

    def helper(self, left, right, n, current, res):
        if left == right and right == n:
            res.append(current)
            return
        if n > left:
            self.helper(left+1, right, n, current+'(', res)
        if right < left:
            self.helper(left, right+1, n, current+')', res)
        return res

File: problems/generateParentheses.js
/**
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
] 
*/

/**
 * Classic DFS, two possible NEXT MOVES are (1) Add left and (2) Add right base on different conditions 
 */

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    
  const res = [];
  dfs(0, 0, "");
  
  return res;
  
  function dfs(right, left, current) {
      if(left === right && right === n) {
          res.push(current);
          return;
      }
      if(left < n) {
          dfs(right, left+1, current+"(");
      }
      if(right < left) {
          dfs(right+1, left, current+")");
      }
  }
};